### DISCUSSION TITLE
TcpClient, TcpServer, UART

### DISCUSSION THREAD

**Discussion by WiilyLin** [nasa/fprime#2281] [General]
What is different for this 3 ? Because there are no this selection in the past? What is corresponding function in the past?

Select communication driver type:
1 - TcpClient
2 - TcpServer
3 - UART
Choose from 1, 2, 3 [1]:

**Comment by thomas-bc** [comment]
@WiilyLin this is a new feature!

This will change the type of the comDriver component.
Selecting the default option (1 - TcpClient) will keep the exact same structure as before. TcpServer and UART will generate a deployment that uses a communication driver that uses these protocols. For example if you are running on a microcontroller and want to connect a GDS through UART instead of TCP, you can use that 3rd option.

Does this help?

**Reply by WiilyLin** [reply]
Of course. At least I know that there are same as before when I use TcpClient. But I'm not very sure in which situation I will use TcpServer. Is it also can build in microcontroller?

**Reply by thomas-bc** [reply]
TcpServer will be useful if you have firewall rules on the machine that runs the GDS that prevents FSW to initiate the connection. For example if you're running on WSL on Windows, that can help resolve some issues. If you haven't noticed any connectivity issues in your testing, I wouldn't worry about it for now. It can always be switched in the future. If you want to see how that works, the LedBlinker tutorial uses TcpServer I believe.

**Reply by WiilyLin** [reply]
So if we want to run flight software on the virtual machine. Maybe TCPserver is the better choice to use. Because we also want to use some simulators, like NOS3 engine.
